experimental/dds/tree2/src/feature-libraries/chunked-forest/chunkedForest.ts

```diff
@@ -142,43 +137,62 @@ class ChunkedForest extends SimpleDependee implements IEditableForest {
 				this.forest.activeVisitor = undefined;
 				this.forest.events.emit("afterChange");
 			},
-			onDelete(index: number, count: number): void {
-				this.onMoveOut(index, count);
+			destroy(range: Range): void {
+				this.detachEdit(range, undefined);
 			},
-			onInsert(index: number, content: Delta.ProtoNodes): void {
+			create(index: PlaceIndex, content: Delta.ProtoNodes): void {
 				this.forest.invalidateDependents();
 				const chunks: TreeChunk[] = content.map((c) => chunkTree(c, this.forest.chunker));
-				const field = this.forest.newDetachedField();
-				this.forest.roots.fields.set(detachedFieldAsKey(field), chunks);
-				this.moveIn(index, field, false);
+				const parent = this.getParent();
+				this.forest.roots.fields.set(parent.key, chunks);
+			},
+			attach(source: DetachedRangeUpPath, destination: PlaceIndex): void {
+				this.forest.invalidateDependents();
+				this.attachEdit(source, destination);
 			},
-			onMoveOut(index: number, count: number, id?: Delta.MoveId): void {
+			detach(source: Range, destination: DetachedPlaceUpPath): void {
 				this.forest.invalidateDependents();
+				this.detachEdit(source, destination);
+			},
+			attachEdit(source: DetachedRangeUpPath, destination: PlaceIndex): void {
+				const sourceField = this.forest.roots.fields.get(source.field) ?? [];
+				assertValidIndex(source.start, sourceField, true);
+				assertValidIndex(source.end, sourceField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+
+				this.moveIn(destination, source.field);
+			},
+			detachEdit(source: Range, destination: DetachedPlaceUpPath | undefined): void {
 				const parent = this.getParent();
 				const sourceField = parent.mutableChunk.fields.get(parent.key) ?? [];
-				const newField = sourceField.splice(index, count);
 
-				if (id !== undefined) {
-					const detached = this.forest.newDetachedField();
-					const key = detachedFieldAsKey(detached);
+				assertValidIndex(source.start, sourceField, true);
+				assertValidIndex(source.end, sourceField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+				const newField = sourceField.splice(source.start, source.end - source.start);
+
+				if (sourceField.length === 0) {
```

Comment:
moving this up from where it was makes the case where source and destination match incorrect. Object forest has same issue. Either explicitly forbid this or put it back so that case works.

Reply:
adding a comment to it to clarify why its at the end, and maybe a unit test might be a good idea to prevent future similar issues

Reply:
this check should still be moved to the end: its just safer to have it there and there is no reason not to.